From ba667e47ce08de2e8314ab8b6cc1f916152d44a8 Mon Sep 17 00:00:00 2001
From: Nathan Chancellor <nathan@kernel.org>
Date: Thu, 2 Jun 2022 08:31:03 -0700
Subject: [PATCH 11/18] NOT-CBL: Revert "workqueue: Wrap flush_workqueue()
 using a macro"

This reverts commit 3b8d1c0d6306eb55b463ef93db12ed472fc68bc5.

The warnings should be fixed first.

Signed-off-by: Nathan Chancellor <nathan@kernel.org>
---
 include/linux/workqueue.h | 64 +++++----------------------------------
 kernel/workqueue.c        | 16 +++-------
 2 files changed, 12 insertions(+), 68 deletions(-)

diff --git a/include/linux/workqueue.h b/include/linux/workqueue.h
index e1f1c8b1121b..7fee9b6cfede 100644
--- a/include/linux/workqueue.h
+++ b/include/linux/workqueue.h
@@ -445,7 +445,7 @@ extern bool mod_delayed_work_on(int cpu, struct workqueue_struct *wq,
 			struct delayed_work *dwork, unsigned long delay);
 extern bool queue_rcu_work(struct workqueue_struct *wq, struct rcu_work *rwork);
 
-extern void __flush_workqueue(struct workqueue_struct *wq);
+extern void flush_workqueue(struct workqueue_struct *wq);
 extern void drain_workqueue(struct workqueue_struct *wq);
 
 extern int schedule_on_each_cpu(work_func_t func);
@@ -563,23 +563,15 @@ static inline bool schedule_work(struct work_struct *work)
 	return queue_work(system_wq, work);
 }
 
-/*
- * Detect attempt to flush system-wide workqueues at compile time when possible.
- *
- * See https://lkml.kernel.org/r/49925af7-78a8-a3dd-bce6-cfc02e1a9236@I-love.SAKURA.ne.jp
- * for reasons and steps for converting system-wide workqueues into local workqueues.
- */
-extern void __warn_flushing_systemwide_wq(void)
-	__compiletime_warning("Please avoid flushing system-wide workqueues.");
-
 /**
  * flush_scheduled_work - ensure that any scheduled work has run to completion.
  *
  * Forces execution of the kernel-global workqueue and blocks until its
  * completion.
  *
- * It's very easy to get into trouble if you don't take great care.
- * Either of the following situations will lead to deadlock:
+ * Think twice before calling this function!  It's very easy to get into
+ * trouble if you don't take great care.  Either of the following situations
+ * will lead to deadlock:
  *
  *	One of the work items currently on the workqueue needs to acquire
  *	a lock held by your code or its caller.
@@ -594,51 +586,11 @@ extern void __warn_flushing_systemwide_wq(void)
  * need to know that a particular work item isn't queued and isn't running.
  * In such cases you should use cancel_delayed_work_sync() or
  * cancel_work_sync() instead.
- *
- * Please stop calling this function! A conversion to stop flushing system-wide
- * workqueues is in progress. This function will be removed after all in-tree
- * users stopped calling this function.
- */
-/*
- * The background of commit 771c035372a036f8 ("deprecate the
- * '__deprecated' attribute warnings entirely and for good") is that,
- * since Linus builds all modules between every single pull he does,
- * the standard kernel build needs to be _clean_ in order to be able to
- * notice when new problems happen. Therefore, don't emit warning while
- * there are in-tree users.
  */
-#define flush_scheduled_work()						\
-({									\
-	if (0)								\
-		__warn_flushing_systemwide_wq();			\
-	__flush_workqueue(system_wq);					\
-})
-
-/*
- * Although there is no longer in-tree caller, for now just emit warning
- * in order to give out-of-tree callers time to update.
- */
-#define flush_workqueue(wq)						\
-({									\
-	struct workqueue_struct *_wq = (wq);				\
-									\
-	if ((__builtin_constant_p(_wq == system_wq) &&			\
-	     _wq == system_wq) ||					\
-	    (__builtin_constant_p(_wq == system_highpri_wq) &&		\
-	     _wq == system_highpri_wq) ||				\
-	    (__builtin_constant_p(_wq == system_long_wq) &&		\
-	     _wq == system_long_wq) ||					\
-	    (__builtin_constant_p(_wq == system_unbound_wq) &&		\
-	     _wq == system_unbound_wq) ||				\
-	    (__builtin_constant_p(_wq == system_freezable_wq) &&	\
-	     _wq == system_freezable_wq) ||				\
-	    (__builtin_constant_p(_wq == system_power_efficient_wq) &&	\
-	     _wq == system_power_efficient_wq) ||			\
-	    (__builtin_constant_p(_wq == system_freezable_power_efficient_wq) && \
-	     _wq == system_freezable_power_efficient_wq))		\
-		__warn_flushing_systemwide_wq();			\
-	__flush_workqueue(_wq);						\
-})
+static inline void flush_scheduled_work(void)
+{
+	flush_workqueue(system_wq);
+}
 
 /**
  * schedule_delayed_work_on - queue work in global workqueue on CPU after delay
diff --git a/kernel/workqueue.c b/kernel/workqueue.c
index 1ea50f6be843..4056f2a3f9d5 100644
--- a/kernel/workqueue.c
+++ b/kernel/workqueue.c
@@ -2788,13 +2788,13 @@ static bool flush_workqueue_prep_pwqs(struct workqueue_struct *wq,
 }
 
 /**
- * __flush_workqueue - ensure that any scheduled work has run to completion.
+ * flush_workqueue - ensure that any scheduled work has run to completion.
  * @wq: workqueue to flush
  *
  * This function sleeps until all work items which were queued on entry
  * have finished execution, but it is not livelocked by new incoming ones.
  */
-void __flush_workqueue(struct workqueue_struct *wq)
+void flush_workqueue(struct workqueue_struct *wq)
 {
 	struct wq_flusher this_flusher = {
 		.list = LIST_HEAD_INIT(this_flusher.list),
@@ -2943,7 +2943,7 @@ void __flush_workqueue(struct workqueue_struct *wq)
 out_unlock:
 	mutex_unlock(&wq->mutex);
 }
-EXPORT_SYMBOL(__flush_workqueue);
+EXPORT_SYMBOL(flush_workqueue);
 
 /**
  * drain_workqueue - drain a workqueue
@@ -2971,7 +2971,7 @@ void drain_workqueue(struct workqueue_struct *wq)
 		wq->flags |= __WQ_DRAINING;
 	mutex_unlock(&wq->mutex);
 reflush:
-	__flush_workqueue(wq);
+	flush_workqueue(wq);
 
 	mutex_lock(&wq->mutex);
 
@@ -6111,11 +6111,3 @@ void __init workqueue_init(void)
 	wq_online = true;
 	wq_watchdog_init();
 }
-
-/*
- * Despite the naming, this is a no-op function which is here only for avoiding
- * link error. Since compile-time warning may fail to catch, we will need to
- * emit run-time warning from __flush_workqueue().
- */
-void __warn_flushing_systemwide_wq(void) { }
-EXPORT_SYMBOL(__warn_flushing_systemwide_wq);
-- 
2.36.1

